Introduce MutableX

lchanmann opened the pull request:
This happens for context that is a proxy of OrX, when calling put the OrX was replaced by a new OrX and the proxyX carries on with the new OrX.

This manifested when sudo-ing inside rule action outside of agency.submit.
- Before sudo-ing there is a subject in the context and spid=intuit
<img width="678" alt="Screen Shot 2022-10-06 at 5 01 04 PM" src="https://user-images.githubusercontent.com/441658/194425478-f1b13bc9-e997-438f-bf07-a78fd351d662.png">

- after sudo-ing, the original context subject and spid were cleared
<img width="846" alt="Screen Shot 2022-10-06 at 5 02 06 PM" src="https://user-images.githubusercontent.com/441658/194425475-e06a8ffd-b92c-4ddf-bd5b-82ee06f129a7.png">

lchanmann committed:
Prevent resetting the original context

this happens for context that is a proxy of OrX, when calling put the OrX was replaced by a new OrX and the proxyX carries on with the new OrX

jlhughes commented on src/foam/nanos/session/Session.js:
```diff
@@ -261,7 +261,7 @@ List entries are of the form: 172.0.0.0/24 - this would restrict logins to the 1
       javaCode: `
       Subject subject = new Subject.Builder(x).setUser(null).build();
-        return x
+        return new OrX(x)
```
I don't understand this. The point of reset is removal or hiding of properties.

jlhughes reviewed (commented):
If reset is the issue, then the solution is to not reset, rather than wrap in OrX as that simply nullifies the effects of reset. 
we need to understand the scenarios where reset is required.

jlhughes commented:
@lchanmann in your example, are you sure the email resolves to a user? 
sudo(email) -> sudo(user) 
which creates the session and sets the user id. 
in session.applyTo, a new subject is setup using the userId on the session.

lchanmann committed:
initial MutableX

lchanmann committed:
Revert "Prevent resetting the original context"

This reverts commit 5cdcf7e3993dffa3b72cfc4a0f036192a37818a9.

lchanmann committed:
Fix tests

lchanmann committed:
Fix null effectiveUser throwing "Unknown User" exception

stacktrace:
2022-10-07T14:08:15.444-0400,qtp699221219-187,ERROR,UserCapabilityUpdateInterceptDAO,java.lang.RuntimeException: Unknown user
	at foam.util.Auth.sudo(Auth.java:60)
	at foam.nanos.crunch.ServerCrunchService.updateJunctionFor(ServerCrunchService.java:462)
	at foam.nanos.crunch.UserCapabilityUpdateInterceptDAO.put_(UserCapabilityUpdateInterceptDAO.java:111)
	at foam.dao.ProxyDAO.put_(ProxyDAO.java:466)
	at foam.nanos.approval.ApprovableAwareDAO.put_(ApprovableAwareDAO.java:889)
	at foam.dao.ProxyDAO.put_(ProxyDAO.java:466)
	at foam.nanos.ruler.RulerDAO.put_(RulerDAO.java:2051)
	at foam.dao.ProxyDAO.put_(ProxyDAO.java:466)
	at foam.nanos.auth.CreatedAwareDAO.put_(CreatedAwareDAO.java:44)
	at foam.dao.ProxyDAO.put_(ProxyDAO.java:466)
	at foam.nanos.auth.CreatedByAwareDAO.put_(CreatedByAwareDAO.java:46)
	at foam.dao.ProxyDAO.put_(ProxyDAO.java:466)
	at foam.nanos.auth.LastModifiedAwareDAO.put_(LastModifiedAwareDAO.java:43)
	at foam.dao.ProxyDAO.put_(ProxyDAO.java:466)
	at foam.nanos.auth.LastModifiedByAwareDAO.put_(LastModifiedByAwareDAO.java:48)
	at foam.dao.ProxyDAO.put_(ProxyDAO.java:466)
	at foam.nanos.auth.AuthorizationDAO.put_(AuthorizationDAO.java:283)
	at foam.dao.ProxyDAO.put_(ProxyDAO.java:466)
	at foam.nanos.auth.PermissionedPropertyDAO.put_(PermissionedPropertyDAO.java:121)
	at foam.dao.ProxyDAO.put_(ProxyDAO.java:466)
	at foam.dao.ProxyDAO.put_(ProxyDAO.java:466)
	at foam.dao.ProxyDAO.put_(ProxyDAO.java:466)
	at foam.dao.ProxyDAO.put_(ProxyDAO.java:466)
	at foam.dao.AbstractDAO.put(AbstractDAO.java:674)
	at foam.nanos.dig.drivers.DigFormatDriver.daoPut(DigFormatDriver.java:704)
	at foam.nanos.dig.drivers.DigFormatDriver.put(DigFormatDriver.java:530)
	at foam.nanos.dig.DigWebAgent.execute(DigWebAgent.java:59)
	at foam.nanos.http.HttpParametersWebAgent.execute(HttpParametersWebAgent.java:234)
	at foam.nanos.pm.PMWebAgent.execute(PMWebAgent.java:34)
	at foam.nanos.http.ProxyWebAgent.execute(ProxyWebAgent.java:193)
	at foam.nanos.http.AuthWebAgent.execute(AuthWebAgent.java:89)
	at foam.nanos.http.NanoRouter.service(NanoRouter.java:127)

kgrgreer commented:
Beautiful
